Generic type inference prefers conditional types

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.4.5 and 3.5.1

<!-- Search terms you tried before logging this (so others can find this issue more easily) -->
**Search Terms:** Conditional types, Type inference

**Code**

```ts
interface I {a: string}
type Cond<T> = T extends any ? Partial<T>: T;

declare function f<T>(x1: T, x2: Cond<T>): T;
function g(p1: I, p2: Cond<I>): I {
  return f(p1, p2); // error here
}
```

**Expected behavior:**
No errors. A trivial type inference is `T=I`. This is also the behavior of TypeScript 3.3.

**Actual behavior:**
`error TS2322: Type 'Partial<I>' is not assignable to type 'I'.`
From that error message the actual inference is `T=Partial<I>` (or `Cond<I>`), which seems like a regression to me.

**Playground Link:** <!-- A link to a TypeScript Playground "Share" link which demonstrates this behavior -->
[playground link](https://www.typescriptlang.org/play/#src=interface%20I%20%7Ba%3A%20string%7D%0D%0Atype%20Cond%3CT%3E%20%3D%20T%20extends%20any%20%3F%20Partial%3CT%3E%3A%20T%3B%0D%0A%0D%0Adeclare%20function%20f%3CT%3E(x1%3A%20T%2C%20x2%3A%20Cond%3CT%3E)%3A%20T%3B%0D%0Afunction%20g(p1%3A%20I%2C%20p2%3A%20Cond%3CI%3E)%3A%20I%20%7B%0D%0A%20%20return%20f(p1%2C%20p2)%3B%20%2F%2F%20error%20here%0D%0A%7D)

**Related Issues:** <!-- Did you find other bugs that looked similar? -->
There are some issues around conditional types but nothing quite a duplication, I think.
https://github.com/microsoft/TypeScript/issues/31601 requires function overloads
https://github.com/microsoft/TypeScript/issues/24085 is a problem with enums
https://github.com/microsoft/TypeScript/issues/30489 is more about type inference on the conditional type itself, not the generic type when the conditional type is used
https://github.com/microsoft/TypeScript/issues/30341 is similar but it's marked as fixed

@rkirov

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -15354,36 +15354,11 @@ namespace ts {
                     inferFromTypes(getFalseTypeFromConditionalType(<ConditionalType>source), getFalseTypeFromConditionalType(<ConditionalType>target));
                 }
                 else if (target.flags & TypeFlags.Conditional && !contravariant) {
-                    inferFromTypes(source, getTrueTypeFromConditionalType(<ConditionalType>target));
-                    inferFromTypes(source, getFalseTypeFromConditionalType(<ConditionalType>target));
+                    const targetTypes = [getTrueTypeFromConditionalType(<ConditionalType>target), getFalseTypeFromConditionalType(<ConditionalType>target)];
+                    inferToMultipleTypes(source, targetTypes, /*isIntersection*/ false);
                 }
                 else if (target.flags & TypeFlags.UnionOrIntersection) {
-                    // We infer from types that are not naked type variables first so that inferences we
-                    // make from nested naked type variables and given slightly higher priority by virtue
-                    // of being first in the candidates array.
-                    let typeVariableCount = 0;
-                    for (const t of (<UnionOrIntersectionType>target).types) {
-                        if (getInferenceInfoForType(t)) {
-                            typeVariableCount++;
-                        }
-                        else {
-                            inferFromTypes(source, t);
-                        }
-                    }
-                    // Inferences directly to naked type variables are given lower priority as they are
-                    // less specific. For example, when inferring from Promise<string> to T | Promise<T>,
-                    // we want to infer string for T, not Promise<string> | string. For intersection types
-                    // we only infer to single naked type variables.
-                    if (target.flags & TypeFlags.Union ? typeVariableCount !== 0 : typeVariableCount === 1) {
-                        const savePriority = priority;
-                        priority |= InferencePriority.NakedTypeVariable;
-                        for (const t of (<UnionOrIntersectionType>target).types) {
-                            if (getInferenceInfoForType(t)) {
-                                inferFromTypes(source, t);
-                            }
-                        }
-                        priority = savePriority;
-                    }
+                    inferToMultipleTypes(source, (<UnionOrIntersectionType>target).types, !!(target.flags & TypeFlags.Intersection));
                 }
                 else if (source.flags & TypeFlags.Union) {
                     // Source is a union or intersection type, infer from each constituent type
@@ -15479,6 +15454,35 @@ namespace ts {
                     }
                 }
                 return undefined;
+            }
+
+            function inferToMultipleTypes(source: Type, targets: Type[], isIntersection: boolean) {
+                // We infer from types that are not naked type variables first so that inferences we
+                // make from nested naked type variables and given slightly higher priority by virtue
+                // of being first in the candidates array.
+                let typeVariableCount = 0;
+                for (const t of targets) {
+                    if (getInferenceInfoForType(t)) {
+                        typeVariableCount++;
+                    }
+                    else {
+                        inferFromTypes(source, t);
+                    }
+                }
+                // Inferences directly to naked type variables are given lower priority as they are
+                // less specific. For example, when inferring from Promise<string> to T | Promise<T>,
+                // we want to infer string for T, not Promise<string> | string. For intersection types
+                // we only infer to single naked type variables.
+                if (isIntersection ? typeVariableCount === 1 : typeVariableCount !== 0) {
+                    const savePriority = priority;
+                    priority |= InferencePriority.NakedTypeVariable;
+                    for (const t of targets) {
+                        if (getInferenceInfoForType(t)) {
+                            inferFromTypes(source, t);
+                        }
+                    }
+                    priority = savePriority;
+                }
             }
 
             function inferToMappedType(source: Type, target: MappedType, constraintType: Type): boolean {

--- a/tests/baselines/reference/conditionalTypeRelaxingConstraintAssignability.js
+++ b/tests/baselines/reference/conditionalTypeRelaxingConstraintAssignability.js
@@ -23,6 +23,20 @@ new Elem('' as ElChildren.Text);
 new Elem('' as ElChildren.Void | ElChildren.Text); // error
 new Elem('' as ElChildren); // error
 
+// Repro from #31766
+
+interface I { a: string }
+
+type DeepPartial<T> =
+    T extends object ? {[K in keyof T]?: DeepPartial<T[K]>} : T;
+
+declare function f<T>(t: T, partial: DeepPartial<T>): T;
+
+function g(p1: I, p2: Partial<I>): I {
+  return f(p1, p2);
+}
+
+
 //// [conditionalTypeRelaxingConstraintAssignability.js]
 "use strict";
 exports.__esModule = true;
@@ -37,3 +51,6 @@ new Elem(undefined);
 new Elem('');
 new Elem(''); // error
 new Elem(''); // error
+function g(p1, p2) {
+    return f(p1, p2);
+}

--- a/tests/baselines/reference/conditionalTypeRelaxingConstraintAssignability.symbols
+++ b/tests/baselines/reference/conditionalTypeRelaxingConstraintAssignability.symbols
@@ -71,3 +71,47 @@ new Elem('' as ElChildren); // error
 >Elem : Symbol(Elem, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 8, 83))
 >ElChildren : Symbol(ElChildren, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 0, 0), Decl(conditionalTypeRelaxingConstraintAssignability.ts, 2, 20))
 
+// Repro from #31766
+
+interface I { a: string }
+>I : Symbol(I, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 22, 27))
+>a : Symbol(I.a, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 26, 13))
+
+type DeepPartial<T> =
+>DeepPartial : Symbol(DeepPartial, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 26, 25))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 28, 17))
+
+    T extends object ? {[K in keyof T]?: DeepPartial<T[K]>} : T;
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 28, 17))
+>K : Symbol(K, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 29, 25))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 28, 17))
+>DeepPartial : Symbol(DeepPartial, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 26, 25))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 28, 17))
+>K : Symbol(K, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 29, 25))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 28, 17))
+
+declare function f<T>(t: T, partial: DeepPartial<T>): T;
+>f : Symbol(f, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 29, 64))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 31, 19))
+>t : Symbol(t, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 31, 22))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 31, 19))
+>partial : Symbol(partial, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 31, 27))
+>DeepPartial : Symbol(DeepPartial, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 26, 25))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 31, 19))
+>T : Symbol(T, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 31, 19))
+
+function g(p1: I, p2: Partial<I>): I {
+>g : Symbol(g, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 31, 56))
+>p1 : Symbol(p1, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 33, 11))
+>I : Symbol(I, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 22, 27))
+>p2 : Symbol(p2, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 33, 17))
+>Partial : Symbol(Partial, Decl(lib.es5.d.ts, --, --))
+>I : Symbol(I, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 22, 27))
+>I : Symbol(I, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 22, 27))
+
+  return f(p1, p2);
+>f : Symbol(f, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 29, 64))
+>p1 : Symbol(p1, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 33, 11))
+>p2 : Symbol(p2, Decl(conditionalTypeRelaxingConstraintAssignability.ts, 33, 17))
+}
+

--- a/tests/baselines/reference/conditionalTypeRelaxingConstraintAssignability.types
+++ b/tests/baselines/reference/conditionalTypeRelaxingConstraintAssignability.types
@@ -62,3 +62,30 @@ new Elem('' as ElChildren); // error
 >'' as ElChildren : ElChildren
 >'' : ""
 
+// Repro from #31766
+
+interface I { a: string }
+>a : string
+
+type DeepPartial<T> =
+>DeepPartial : DeepPartial<T>
+
+    T extends object ? {[K in keyof T]?: DeepPartial<T[K]>} : T;
+
+declare function f<T>(t: T, partial: DeepPartial<T>): T;
+>f : <T>(t: T, partial: DeepPartial<T>) => T
+>t : T
+>partial : DeepPartial<T>
+
+function g(p1: I, p2: Partial<I>): I {
+>g : (p1: I, p2: Partial<I>) => I
+>p1 : I
+>p2 : Partial<I>
+
+  return f(p1, p2);
+>f(p1, p2) : I
+>f : <T>(t: T, partial: DeepPartial<T>) => T
+>p1 : I
+>p2 : Partial<I>
+}
+

--- a/tests/cases/compiler/conditionalTypeRelaxingConstraintAssignability.ts
+++ b/tests/cases/compiler/conditionalTypeRelaxingConstraintAssignability.ts
@@ -22,3 +22,16 @@ new Elem(undefined as ElChildren.Void);
 new Elem('' as ElChildren.Text);
 new Elem('' as ElChildren.Void | ElChildren.Text); // error
 new Elem('' as ElChildren); // error
+
+// Repro from #31766
+
+interface I { a: string }
+
+type DeepPartial<T> =
+    T extends object ? {[K in keyof T]?: DeepPartial<T[K]>} : T;
+
+declare function f<T>(t: T, partial: DeepPartial<T>): T;
+
+function g(p1: I, p2: Partial<I>): I {
+  return f(p1, p2);
+}

